RFC: create state.exists

mchugh19 opened the pull request:
### What does this PR do?
A hack which duplicates the state.show_sls to return True or False. Because states can live in gitfs, a simple file.exists isn't enough to support this need.

See this top file where states are assigned from a roles grain. 
States are assigned to hosts, based on the roles listed in the roles grain. Without any other check, any roles in the grain but missing accompanying state files, will cause highstate to fail. By wrapping the role assignment in top with a check to validate it exists, these failures can be avoided.

```
# cat top.sls
base:
  '*':
    - common

  {% set roles = salt['grains.get']('roles',[]) -%}
  {% for role in roles -%}
  {% if salt['state.exists']('roles.{0}'.format(role)) -%}
  'roles:{{ role }}':
    - match: grain
    - roles.{{ role }}
  {% endif -%}
  {% endfor -%}

# salt-call state.show_top
[ERROR   ] Template was specified incorrectly: False
local:
    ----------
    base:
        - common
# touch roles/saltmaster.sls
# salt-call state.show_top
local:
    ----------
    base:
        - common
        - roles.saltmaster

```

### What issues does this PR fix or reference?
Intent was to cover use cases such as #16687 
It would be better to add functionality like ignore_missing to states, but how to do so was not immediately obvious. 

### Tests written?
No

### Commits signed with GPG?
No

Commit:
create state.exists

Commit:
Merge branch 'develop' into state.exists

terminalmage reviewed (changes requested):
Duplicating all of the code in `show_sls` is not necessary here. If there were errors, a list is returned, if not, a dict is returned. Therefore, this function can re-written like so:

```python
return isinstance(
    show_sls(mods, test=test, queue=queue, **kwargs),
    dict
)
```

Also, I would prefer the name of the function be changed to `sls_exists`. This leaves open the option later on of adding a function to check if a given ID exists within a specified SLS file, and naming that function `id_exists`.

Commit:
refine state_exists and add id_exists

Commit:
Merge branch 'state.exists' of github.com:mchugh19/salt into state.exists

Commit:
Handle invalid states on id_exists
Add unit tests

mchugh19 commented:
Fixed and renamed to sls_exists. Also added id_exists, and unit tests for both.

terminalmage commented on salt/modules/state.py:
```diff
@@ -1880,6 +1880,59 @@ def show_sls(mods, test=None, queue=False, **kwargs):
 
+def sls_exists(mods, test=None, queue=False, **kwargs):
+    '''
+    Tests for the existance the of a specific sls or list of sls files on the
```
SLS should be all-capitals when referring to an SLS file.

mchugh19 replied:
Corrected

terminalmage reviewed (changes requested):
A few docstring changes need to be made, and the `id_exists` function doesn't quite do what I was describing earlier. I've left an explanation and some example code to get you pointed in the right direction.

mchugh19 commented on salt/modules/state.py:
```diff
@@ -1880,6 +1880,59 @@ def show_sls(mods, test=None, queue=False, **kwargs):
+    for state_id in ids:
+        if state_id not in state_ids:
+            return False
+    return True
```
I believe the described functionality is what was implemented. The line 
```state_ids = show_sls(mods, test=test, queue=queue, **kwargs).keys()```
causes the state to be rendered and the top level key names gathered. Thus, the state_ids variable should be a list of the state IDs. 

Sounds like using __id__ key from lowstate might be a superior method, so I'll take a look.

mchugh19 replied:
Done

Commit:
Fix wording
Improve id_exists by using show_low_sls and checking __id__ keys.

mchugh19 commented on salt/modules/state.py:
```diff
@@ -1880,6 +1880,59 @@ def show_sls(mods, test=None, queue=False, **kwargs):
+def sls_exists(mods, test=None, queue=False, **kwargs):
+    '''
+    Tests for the existance the of a specific sls or list of sls files on the
+    master. Similar to show_sls, rather than returning state details, returns
```
Done

mchugh19 commented on salt/modules/state.py:
```diff
@@ -1880,6 +1880,59 @@ def show_sls(mods, test=None, queue=False, **kwargs):
+    .. versionadded:: Fluorine
+
+    saltenv
+        Specify a salt fileserver environment to be used when applying states
```
Corrected

mchugh19 commented on salt/modules/state.py:
```diff
@@ -1880,6 +1880,59 @@ def show_sls(mods, test=None, queue=False, **kwargs):
+    '''
+    Tests the for a specific id or list of ids within a given state on the
+    master. Similar to sls_exists, returns True or False. The default
+    environment is ``base``, use ``saltenv`` to specify a different environment.
```
Done

mchugh19 commented on salt/modules/state.py:
```diff
@@ -1880,6 +1880,59 @@ def show_sls(mods, test=None, queue=False, **kwargs):
+    .. versionadded:: Fluorine
+
+    saltenv
+        Specify a salt fileserver environment to be used when applying states
```
Corrected

mchugh19 commented:
Thanks @terminalmage. Suggestions/corrections applied.

mchugh19 commented:
I think all of the requested updates have been submitted. Is there anything outstanding?

terminalmage reviewed (approved):
This looks good, thanks for your patience as we've worked through this.

isbm commented on salt/modules/state.py:
```diff
@@ -1880,6 +1880,57 @@ def show_sls(mods, test=None, queue=False, **kwargs):
+        salt '*' state.id_exists create_myfile,update_template filestate saltenv=dev
+    '''
+    if isinstance(ids, six.string_types):
+        ids = ids.split(',')
```
Please use `salt.utils.args.split_input` instead of this:
```python
def id_exists(...
    sls_ids = set(x['__id__'] for x ...
    return set(salt.utils.args.split_input(ids)).issubset(sls_ids)
```
And it would be nice to others not to use one-char variables.

terminalmage replied:
I don't mind the 1-char in a comprehension expression, to be honest. I don't think that needs to change.

@isbm is right about the split_input, we do have a helper for this. It could be done as

```python
ids = salt.utils.args.split_input(ids)
```

Commit:
convert instances of variable.split(',') to use salt.utils.args.split_input(variable)

mchugh19 commented:
Updated to use salt.utils.args.split_input. While there, I also updated the other uses of variable.split(',') in the file to keep everything consistent. Is that alright?

isbm commented:
> Updated to use salt.utils.args.split_input. While there, I also updated the other uses of variable.split(',') in the file to keep everything consistent. Is that alright?

Yes, this is alright. Almost. :wink: You just don't need to extra-check `if isinstance...` is string. Just split directly (and so remove all those checks by the way).

Thanks!

Commit:
properly use salt.utils.args.split and remove wrapping logic

mchugh19 commented:
Ah, that makes sense. Done!

isbm reviewed (approved):
@mchugh19 @rallytime @terminalmage LGTM! :+1:

Commit:
Merge branch 'develop' into state.exists
